---
title: "Trait-based community assembly and succession in the infant gut microbiome."
author: "John Guittar, Ashley Shade, Elena Litchman"
output:
  pdf_document: default
  html_document:
    df_print: paged
---

\begin{center}
\bigskip
\bigskip
\bigskip
\bigskip
{\Huge Tables and figures}
\end{center}

\pagebreak


```{r global_options, include = FALSE}

knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE, cache = TRUE)

```


```{r load data, include = FALSE}

##Note... for which analysis are otus from 'unclassified' genera valid? Should I give these unkonwn genera specific keys? What about when I have trait data for an otu ID with an 'unclassified' genus -- maybe I could propagate trait data among different 'unclassified' genera, if they have some trait info.
wd <- 'C:\\Users\\John\\Documents\\msu\\microbiome_trait_succession\\'
setwd(wd)
source('custom_functions.R')
loadpax(pkg = c('gtable','tidyverse','knitr','data.table','vegan','GGally','grid','gridExtra','betapart','scales','stringr', 'kableExtra'))

# load data

# Should we use abundance data normalized by 16S copy number?
# https://microbiomejournal.biomedcentral.com/articles/10.1186/s40168-018-0420-9
# Yes, because we use results from bugbase, which are already normalized
# otus_wide <- readRDS('data\\otus_wide.RDS')
otus_wide <- readRDS('data\\otus_wide_normalized.RDS') %>%
  mutate_if(is.factor, as.character)
meta <- readRDS('data\\meta.RDS') %>%
  mutate_if(is.factor, as.character)
tax <- readRDS('data\\tax.RDS') %>%
  mutate_if(is.factor, as.character)
traits <- readRDS('data\\traits_castor_subtree_averaging.RDS') %>%
  filter(otu %in% names(otus_wide)) %>%
  spread(trait, val) %>%
  mutate_if(is.factor, as.character)
traits_sparse <- fread('data\\traits_sparse.csv') #before trait inference using castor

#create versions of data with and without c-section infants
#With C-section data, skinny table
otus_cs <- otus_wide %>%
  left_join(meta[, c('subject','delivery')], by = 'subject') %>%
  gather(otu, abun, -sampleID, -subject, -t, -delivery) %>%
  filter(abun > 0)

# Without C-section data, skinny table
otus <- otus_cs %>%
  filter(delivery != 'caesaren') %>%
  select(-delivery)

# Without C-section data, wide table
otus_wide <- spread(otus, otu, abun, fill = 0)

#Set plotting theme
th <- theme_bw() +
  theme(
    panel.grid.minor = element_blank(),
    panel.grid.major = element_blank(),
    panel.background = element_blank())

```


```{r table 1}

# rename columns
t_names <- sort(c(names(traits)[-1]))
             
#renaming
tmp <- c(
  "B_vitamins" = "B vitamins",
  "Copies_16S" = "16S gene copies",
  "Forms_Biofilms" = "Forms biofilms",
  "GC_content" = "GC content",
  "Gene_number" = "Genes",
  "Gram positive" = "Gram_Positive",
  "IgA" = "IgA binding affinity",
  "Temp_optimum" = "Temperature optimum"
)

#trait source list
t_sources <- c("BacDive",
               "Browne et al. 2016",
               "IJSEM",
               "JGI", 
               "Magnusdottir et al. 2015",
               "NCBI",
               "Palm et al. 2014",
               "Stoddard et al. 2015",
               "Ward et al. 2017")

names(t_sources) <- seq_along(t_sources)

# trait sources
tmp <- list(
  "16S gene copies"          = c("Stoddard et al. 2015","Ward et al. 2017"),
  #"Aggregation"              = c("IJSEM"),
  "Obligate anaerobe"        = c("Ward et al. 2017"),
  "B vitamins"               = c("Magnusdottir et al. 2015"),
  #"Facultatively anaerobic"  = c("Ward et al. 2017"),
  "Forms biofilms"           = c("Ward et al. 2017"),
  "GC content"               = c("BacDive","IJSEM"),
  "Genes"                    = c("NCBI"),
  #"Genome size"              = c("NCBI"),
  "Gram positive"            = c("Ward et al. 2017"),
  "IgA binding affinity"     = c("Palm et al. 2014"),  
  "Length"                   = c("BacDive","IJSEM","JGI"),   
  "Motility"                 = c("IJSEM","JGI"),  
  #"pH optimum"               = c("BacDive","IJSEM","JGI"),    
  #"Potentially pathogenic"   = c("Ward et al. 2017"),
  #"Salt optimum"             = c("BacDive","IJSEM"),  
  "Sporulation"              = c("Browne et al. 2016","IJSEM","JGI"),  
  "Temperature optimum"      = c("IJSEM","BacDive"),  
  "Width"                    = c("BacDive","IJSEM","JGI")
)

tmp <- unlist(lapply(tmp, function(x) paste(names(t_sources)[match(x, t_sources)], collapse = ",")))
t_sources <- data.frame(trait = names(tmp), Sources = tmp, row.names = NULL, stringsAsFactors = FALSE)

tmp <- c(
  "16S gene copies"          = "No. in genome",
  "Aggregation"              = "0 or 1",
  "Obligate anaerobe"        = "0 or 1",
  "B vitamins"               = "No. pathways in genome",
  "Facultatively anaerobic"  = "0 or 1",
  "Forms biofilms"           = "0 or 1",
  "GC content"               = "Percent (\\%)",
  "Genes"                    = "No. in genome",
  "Genome size"              = "Mb",
  "Gram positive"            = "0 or 1",
  "IgA binding affinity"     = "log ([IgA+]/[IgA-] + 1)",  
  "Length"                   = "log ($\\mu$m)",   
  "Motility"                 = c("IJSEM","JGI"),  
  "pH optimum"               = "pH",    
  "Potentially pathogenic"   = "0 or 1",
  "Salt optimum"             = "g/l",  
  "Sporulation"              = "Continuous score (0 - 1)",  
  "Temperature optimum"      = "$^{\\circ}$C",  
  "Width"                    = "log ($\\mu$m)"
)

t_units <- data.frame(trait = names(tmp), unit = tmp, row.names = NULL, stringsAsFactors = FALSE)

j <- t_sources %>%
  left_join(t_units, by = 'trait') %>%
  rename(Trait = trait, Units = unit)

kable(j, format = 'latex', escape = F, booktabs = T, align = c('l','l','l','r','r'), linesep = "") %>%
  kable_styling() %>%
  row_spec(0, bold = T) %>% 
  landscape() %>%
  footnote(general = "Sources of trait data used in this study.",
           number = c("BacDive: Bacterial Diversity Metadatabase; ",
                      "Browne et al. 2016; ",
                      "IJSEM: Internat. Jo. of Syst. and Evol. Micro.; ",
                      "JGI: Joint Genome Institute; ",
                      "Magnusdottir et al. 2015; ",
                      "NCBI: the National Center for Biotech. Info.; ",
                      "Palm et al. 2014; ",
                      "Stoddard et al. 2015; ",
                      "Ward et al. 2017."),
           general_title = "Table 1.")


```

\pagebreak

``` {r multipanel succession diagram, fig.width = 8, fig.height = 7}

# Bump up any t<2 to 2 to avoid early noise due to low-sampling.
# Determine centers of colonization periods of OTUs by taking the abundance-weighted mean occurence. 
# Focus only on taxa more than .1% abundant, 
#       only taxa in at least 5 individuals (~10% of analysis population), 
#       only taxa that appear for 3 or more consecutive months.

#after filtering, calculate start, mid, and end times of each otu
j <- otus %>%
  group_by(subject, t) %>%
  filter(abun / sum(abun) > 0.001) %>%
  group_by(otu) %>%
  filter(length(unique(subject)) >= 5) %>%
  mutate(
    t = round(t/30),
    t = ifelse(t < 2, 2, t),
    t = ifelse(t > 36, 36, t)) %>%
  group_by(t, otu) %>%
  summarise(abun = sum(abun)) %>%
  group_by(otu) %>%
  filter(max(t) - min(t) > 3) %>%
  mutate(
    tmean = weighted.mean(t, w = abun),
    start = min(t),
    end = max(t)) %>%
  ungroup()

#calculate percent total community reads, combine with time stats (above)
tmp <- otus %>%
  group_by(otu) %>%
  summarise(abun = sum(abun)) %>%
  ungroup() %>%
  mutate(abun = abun / sum(abun))

j$percent_total_abun <- tmp$abun[match(j$otu, tmp$otu)]

#add rank
#the distinct() is to filter unnecessary dataframe rows
tmp <- j %>%
  ungroup() %>%
  distinct(otu, tmean, .keep_all = TRUE) %>%
  arrange(tmean) %>%
  mutate(otuRank = seq_along(otu))

j$otuRank <- tmp$otuRank[match(j$otu, tmp$otu)]


######### determine succession status using regression coefficients

#note that I divide by the length of unique sampleIDs. This takes care of the off chance that there are multiple samples merged under the 2 or 36 month groups.
#I remove otus that occur in fewer than 5 individuals
#I use otus_wide because the zeros are necessary for the regression
tmp <- otus_wide %>%
  gather(otu, abun, -sampleID, -subject, -t) %>%
  group_by(otu) %>%
  filter(length(unique(subject)) >= 5) %>%
  mutate(
    t = round(t/30),
    t = ifelse(t < 2, 2, t),
    t = ifelse(t > 36, 36, t)) %>%
  group_by(otu, t) %>%
  summarise(abun = sum(abun) / length(unique(paste(sampleID)))) %>%
  do(mod = summary(lm(abun ~ t, .))) %>%
  mutate(
    pval = mod$coefficients[[8]],
    tval = mod$coefficients[[6]],
    group = ifelse(pval < 0.05 & tval < 0, 'Early successional', 'Mid-successional / No trend'),
    group = ifelse(pval < 0.05 & tval > 0, 'Late successional', group),
    group = factor(group, levels = c('Early successional','Mid-successional / No trend','Late successional'))) %>%
  select(otu, group)

# add group designation
j <- j %>%
  left_join(tmp, by = 'otu') %>%
  left_join(tax, by = 'otu') %>%
  mutate(group = factor(group, levels = c('Early successional','Mid-successional / No trend','Late successional')))

p0 <- ggplot(j, aes(x = t, y = otuRank)) +
  geom_line(aes(group = otuRank, color = group), size = 0.75) +
  #geom_point(aes(x = tmean), color = 'black', shape = 1, size = 1, alpha = 0.1) +
  scale_color_brewer(palette = 'Set1') + 
  theme_bw() +
  theme(
    panel.grid.minor = element_blank(),
    panel.grid.major = element_blank(),
    panel.background = element_blank(),
    axis.text.y=element_blank(),
    axis.ticks.y=element_blank(),
    legend.position = 'none') +
  labs(x = 'Month', y = 'OTU, ranked by peak global abundance')

p1 <- j %>%
  filter(group == 'Early successional') %>%
  mutate(abun = abun / sum(abun)) %>%
  group_by(Order, Family) %>%
  summarise(abun = sum(abun)) %>%
  ungroup() %>%
  arrange(-abun) %>%
  mutate(Family = ifelse(Family == 'unclassified', paste(Order, Family, sep = '_'), Family)) %>%
  mutate(Family = ifelse(Family %in% Family[1:5], Family, 'Other')) %>%
  mutate(Family = factor(Family, level = rev(unique(Family)))) %>%
  group_by(Family) %>%
  summarise(abun = sum(abun)) %>%
  ggplot(aes(x = Family, y = abun, fill = Family)) +
   geom_bar(stat = 'identity', color = 'black', fill = '#E41A1C') +
   geom_text(aes(label = Family), hjust = 'left', nudge_y = 0.01) +
   coord_flip() +
   scale_y_continuous(limits = c(0,0.75)) +
   theme_bw() +
   theme(
      panel.grid.minor = element_blank(),
      panel.grid.major = element_blank(),
      panel.background = element_blank(),
      axis.text.y=element_blank(),
      axis.ticks.y=element_blank(),
      legend.position = 'none') +
  labs(y = 'Relative abundance', x = 'Family') +
    ggtitle(" ")

p2 <- j %>%
  filter(group == 'Mid-successional / No trend') %>%
  mutate(abun = abun / sum(abun)) %>%
  group_by(Order, Family) %>%
  summarise(abun = sum(abun)) %>%
  ungroup() %>%
  arrange(-abun) %>%
  mutate(Family = ifelse(Family == 'unclassified', paste(Order, Family, sep = '_'), Family)) %>%
  mutate(Family = ifelse(Family %in% Family[1:5], Family, 'Other')) %>%
  mutate(Family = factor(Family, level = rev(unique(Family)))) %>%
  group_by(Family) %>%
  summarise(abun = sum(abun)) %>%
  ggplot(aes(x = Family, y = abun, fill = Family)) +
   geom_bar(stat = 'identity', color = 'black', fill = '#377EB8') +
   geom_text(aes(label = Family), hjust = 'left', nudge_y = 0.01) +
   coord_flip() +
   scale_y_continuous(limits = c(0,1.05)) +
   theme_bw() +
   theme(
      panel.grid.minor = element_blank(),
      panel.grid.major = element_blank(),
      panel.background = element_blank(),
      axis.text.y=element_blank(),
      axis.ticks.y=element_blank(),
      legend.position = 'none') +
  labs(y = 'Relative abundance', x = 'Family') +
  ggtitle(" ")

p3 <- j %>%
  filter(group == 'Late successional') %>%
  mutate(abun = abun / sum(abun)) %>%
  group_by(Order, Family) %>%
  summarise(abun = sum(abun)) %>%
  ungroup() %>%
  arrange(-abun) %>%
  mutate(Family = ifelse(Family == 'unclassified', paste(Order, Family, sep = '_'), Family)) %>%
  mutate(Family = ifelse(Family %in% Family[1:5], Family, 'Other')) %>%
  mutate(Family = factor(Family, level = rev(unique(Family)))) %>%
  group_by(Family) %>%
  summarise(abun = sum(abun)) %>%
  ggplot(aes(x = Family, y = abun, fill = Family)) +
   geom_bar(stat = 'identity', color = 'black', fill = '#4DAF4A') +
   geom_text(aes(label = Family), hjust = 'left', nudge_y = 0.01) +
   coord_flip() +
   scale_y_continuous(limits = c(0,0.9)) +
   theme_bw() +
   theme(
      panel.grid.minor = element_blank(),
      panel.grid.major = element_blank(),
      panel.background = element_blank(),
      axis.text.y=element_blank(),
      axis.ticks.y=element_blank(),
      legend.position = 'none') +
  labs(y = 'Relative abundance', x = 'Family') +
  ggtitle(" ")

tmp_abuns <- otus %>%
  mutate(
    t = round(t/30),
    t = ifelse(t < 2, 2, t),
    t = ifelse(t > 36, 36, t)) %>%
  group_by(t, otu) %>%
  summarise(abun = sum(abun) / length(unique(sampleID))) %>%
  group_by(t) %>%
  mutate(abun = 100 * abun / sum(abun))
  
tmp <- j %>%
  group_by(t, group) %>%
  summarise(abun = sum(abun)) %>%
  group_by(t) %>%
  mutate(abun = 100 * abun / sum(abun))

ymax <- max(tmp$abun) * 1.1

p4 <- tmp %>%
  filter(group == 'Early successional') %>%
  ggplot(aes(x = t, y = abun)) +
    geom_bar(stat = 'identity', color = '#E41A1C', fill = '#E41A1C', width = 1) +
    scale_y_continuous(limits = c(0, ymax)) +
    th +
    labs(x = 'Month', y = '% Abundance') + 
  ggtitle('Early successional')

p5 <- tmp %>%
  filter(group == 'Mid-successional / No trend') %>%
  ggplot(aes(x = t, y = abun)) +
    geom_bar(stat = 'identity', color = '#377EB8', fill = '#377EB8') +
    scale_y_continuous(limits = c(0,ymax)) +
    th +
    labs(x = 'Month', y = '% Abundance') + 
  ggtitle('Mid-successional / No trend')

p6 <- tmp %>%
  filter(group == 'Late successional') %>%
  ggplot(aes(x = t, y = abun)) +
    geom_bar(stat = 'identity', color = '#4DAF4A', fill = '#4DAF4A') +
    scale_y_continuous(limits = c(0,ymax)) +
    th +
    labs(x = 'Month', y = '% Abundance') + 
  ggtitle('Late successional')

mygrobs <- lapply(list(p0, p3, p2, p1, p6, p5, p4), ggplotGrob)

gt <- gtable(widths=unit(rep(1,3), "null"),
             heights=unit(rep(1,3), "null"))

gt <- gtable_add_grob(gt, mygrobs, 
                       l=c(1,3,3,3,2,2,2),
                       r=c(1,3,3,3,2,2,2),
                       t=c(1,1,2,3,1,2,3),
                       b=c(3,1,2,3,1,2,3))
grid::grid.newpage()
grid::grid.draw(gt)

```

## Figure 1. OTU and family-level abundance patterns over time.
Temporal ranges of taxa over infant gut community development (left), ranked by peak abundance across infants. Peak abundance is calculated as the mean time of OTU presence, weighted by OTU abundance. Each colored row is an OTU, with their x-axis ranges reflecting the first and last times that their mean abundance surpassed 0.1% across all infants. Taxa that never attained 0.1% mean abundance are omitted. OTU colors reflect successional status; taxa were categorized as early (red) or late (green) successional if their global abundances trended significantly (p < 0.05) negative or positive over time based on linear regression; OTUs were categorizeds as mid-successional/no trend (blue) when they did not trend significantly over time. The center column of panels shows percent total abundance of OTUs from each successional category across all infants over time. The right column of panels shows proportions of total community abundance of the top five most abundant families for each successional group.

\pagebreak

```{r traits over time, fig.width = 6, fig.height = 6.5}

otus %>%
  group_by(subject) %>%
  ungroup() %>%
  left_join(traits, by = 'otu') %>%
  gather(trait, val, -sampleID, -subject, -t, -otu, -abun) %>%
  ungroup() %>%
  left_join(meta, by = 'subject') %>%
  group_by(subject, t, trait) %>%
  summarise(val = weighted.mean(val, w = abun, na.rm = TRUE)) %>%
  ungroup() %>%
  mutate(t = round(t/30)) %>%
  filter(t < 37 & t > 1) %>%
  ggplot(aes(x = t, y = val)) +
    stat_summary(fun.y = mean, geom = "point", color = 'red') + 
    stat_summary(fun.data = mean_cl_boot, geom = "linerange", color = 'red') +
    stat_smooth(color = 'black', se = FALSE) +
    facet_wrap(~trait, scale = 'free_y', ncol = 3) +
    labs(x = "Months after birth", y = "Abundance-weighted community mean") +
    theme_bw() +
    theme(
      panel.grid.minor = element_blank(),
      panel.grid.major = element_blank(),
      panel.background = element_blank(),
      legend.position = 'none')


```

## Figure 2. Abundance-weighted trait patterns over time.
Red circles show the average abundance-weighted mean of samples in that month (N = number of samples, and ranges from 26 - 55), and vertical red lines show 95% confidence intervals. Black lines show trends in community weighted trait means over time, calculated using generalized additive models.

\pagebreak


```{r step wise delta diss, fig.height = 5, fig.width = 6}

### bray curtis stuff
meltdist <- function(x, sub, meth = 'bray') {
  # here, make a distance matrix using meth method and then melt it
  x <- as.data.frame(x)
  tz <- x$t
  x$t <- NULL
  x <- x[, sapply(x, is.numeric)]
  row.names(x) <- tz
  x <- vegdist(x, method = meth)
  x <- as.matrix(x)
  x <- melt(x, varnames=c('t1', 't2'), value.name = 'dist')
  x <- data.frame(subject = sub, x)
  x$subject <- as.character(x$subject)
  return(x)
}

#calculate OTU-based dissimilarity
#only examine the window of time when all infants have samples (currently, t_max = 988.2)
j1 <- otus_wide %>%
  filter(t < min(meta$t_max)) %>%
  select(-sampleID) %>%
  group_by(subject) %>%
  do(meltdist(., sub = .$subject[[1]])) %>%
  mutate(group = 'OTU-based dissimilarity')

#how many traits? We will divide euclidean distance by this number to get an average (scaled) distance, so that the abolute distance doesn't change as we add/subtract traits
n_traits <- ncol(traits) - 1

#trait-based dissimilarity - i.e., euclidean distance
j2 <- otus %>%
  left_join(traits, by = 'otu') %>%
  gather(trait, val, -sampleID, -subject, -t, -otu, -abun) %>%
  group_by(sampleID, subject, trait, t) %>%
  summarise(val = weighted.mean(val, w = abun, na.rm = TRUE)) %>%
  filter(t < min(meta$t_max)) %>%
  group_by(trait) %>%
  mutate(val = scale(val)) %>%
  spread(trait, val) %>%
  select(-sampleID) %>%
  group_by(subject) %>%
  do(meltdist(., sub = .$subject[[1]], meth = 'euclidean')) %>%
  mutate(
    dist = dist / n_traits,
    group = 'Trait-based dissimilarity')

#Genus-level taxonomic dissimilarity
#j3 <- otus_wide %>%
#  filter(t < min(meta$t_max)) %>%
#  select(-sampleID) %>%
#  group_by(subject) %>%
#  gather(otu, abun, -subject, -t) %>%
#  left_join(tax, by = 'otu') %>%
#  filter(Genus != 'unclassified') %>%
#  group_by(Genus) %>%
#  mutate(genusID = paste(Genus, as.numeric(as.factor(paste(Order, Family))))) %>%
#  group_by(subject, t, genusID) %>%
#  summarise(abun = sum(abun)) %>%
#  spread(genusID, abun) %>%
#  group_by(subject) %>%
#  do(meltdist(., sub = .$subject[[1]])) %>%
#  mutate(group = 'Taxonomic dissimilarity (genera)')

#combine
j <- bind_rows(j1, j2)

# calculate the dissimilarities between subsequent samples, as a measure of rate of temporal turnover
j1 <- j %>%
  group_by(group, subject, t1) %>%
  filter(t2 > t1) %>%
  filter(t2 == min(t2)) %>%
  mutate(delta = t2 - t1) %>%
  ungroup() %>%
  mutate(group2 = 'Dissimilarity to next sample')

# save for a supplementary figure asking if/how dissimilarity scales with time between samples
j_delta <- j1

# filter to focus on dissimilarity to final sample
j2 <- j %>%
  group_by(group, subject, t1) %>%
  filter(t2 == max(t2) & t1 != max(t2)) %>%
  ungroup() %>%
  mutate(group2 = 'Dissimilarity to final sample')

j <- rbind(select(j1, -delta), j2) %>%
  mutate(group2 = factor(group2, levels = unique(group2)))

j %>%
  filter(t1 <= min(meta$t_max)) %>%
  mutate(t1 = round(t1/30)) %>%
  filter(t1 > 1 & t1 < 32) %>%
  ggplot(aes(x = t1, y = dist)) +
    stat_summary(fun.y = mean, geom = "point", color = 'red') + 
    stat_summary(fun.data = mean_cl_boot, geom = "linerange", color = 'red') +
    stat_smooth(color = 'black', se = FALSE) +
    expand_limits(y = 0) +
    facet_grid(group~group2, scales = 'free') +
    labs(x = 'Months after birth', y = '') +
    th

#mean_cl_boot
#mean_sdl

```

## Figure 3. Taxonomic and trait-based rates of community change within infants.
Red circles show mean dissimilarities between among microbiome samples within infants over their development, grouped by month (N = number of sample comparisons, ranging from 26 - 55), and vertical red lines show 95% confidence intervals. At left, datashow changes in dissimilarity between subsequent samples taken from the same infants over time. At right, lines show dissimilarity between samples and the last sample taken for each infant(up to the 988 day cutoff used to standardize sampling range). Black lines show trends in community weighted trait means over time, calculated using generalized additive models.

\pagebreak

```{r richness over time, fig.height = 3.75}

tmp <- otus %>%
  mutate(t = t / 30) %>%
  filter(t <= 36)

j1 <- tmp %>%
  group_by(subject, t) %>%
  summarise(rich = length(unique(otu)), group1 = 'Instantaneous richness', group2 = 'Within subjects') %>%
  ungroup() %>%
  select(t, subject, rich, group1, group2)

j2 <- tmp %>%
  group_by(t) %>%
  summarise(subject = 1, rich = length(unique(otu)), group1 = 'Instantaneous richness', group2 = 'Across subjects')

j3 <- tmp %>%
  distinct(subject, t, otu) %>%
  group_by(subject) %>%
  arrange(t) %>%
  mutate(rich = cummax(as.numeric(factor(otu, levels = unique(otu))))) %>%
  group_by(subject, t) %>%
  summarise(rich = max(rich), group1 = 'Cumulative richness', group2 = 'Within subjects') %>%
  ungroup() %>%
  select(subject, t, rich, group1, group2)

j4 <- tmp %>%
  distinct(t, otu) %>%
  arrange(t) %>%
  mutate(rich = cummax(as.numeric(factor(otu, levels = unique(otu))))) %>%
  group_by(t) %>%
  summarise(subject = 1, rich = max(rich), group1 = 'Cumulative richness', group2 = 'Across subjects')

j <- rbind(mutate(rbind(j1, j3), scale = 'Individual infants'), 
      mutate(rbind(j2, j4), scale = 'All infants')) %>%
  select(-group2)

j5 <- j %>%
  mutate(t = round(t)) %>%
  group_by(t, group1) %>%
  summarise(
    rich = mean(rich[scale == 'Individual infants']) / mean(rich[scale == 'All infants']),
    scale = 'Individual infants / All infants',
    subject = 1) %>%
  ungroup() 

j <- rbind(j, j5) %>%
  mutate(scale = factor(scale, levels = c('Individual infants', 'All infants', 'Individual infants / All infants')))

j1 <- j %>%
  mutate(
    scale1 = ifelse(scale != 'Individual infants / All infants' & group1 == 'Instantaneous richness', 'Instantaneous richness', 'Individual infants / All infants'),
    scale1 = ifelse(scale != 'Individual infants / All infants' & group1 == 'Cumulative richness', 'Cumulative richness', scale1),
    scale1 = factor(scale1, levels = c('Instantaneous richness','Cumulative richness')))

tmp <- j1 %>%
  filter(scale != 'Individual infants / All infants') %>%
  mutate(t = round(t)) %>%
  filter(t > 1 & t < 36)

ggplot(tmp, aes(x = t, y = rich, color = scale)) +
  stat_summary(fun.y = mean, geom = "point") + 
  stat_summary(fun.data = mean_cl_boot, geom = "linerange", data = filter(tmp, !(scale == 'All infants' & scale1 == 'Cumulative richness'))) +
  stat_smooth(aes(group = interaction(scale, group1)), se = FALSE) +
  scale_color_discrete(name = '') +
  theme_bw() +
   theme(
      panel.grid.minor = element_blank(),
      panel.grid.major = element_blank(),
      panel.background = element_blank(),
      legend.position = 'bottom') +
  facet_wrap(~scale1, scales = 'free_y') +
  labs(x = 'Months after birth', y = 'OTU richness')

```

## Figure 4. Patterns in OTU richness in individual infants over time. 
Colored circles show mean OTU richness of infant microbiome samples over time, grouped by month. Left panel shows instantaneous richness within single infants (red) or across all infants (blue); right panel shows cumulative richness (i.e., the cumulative number of OTUs encountered by one (red) or all (blue) infants over time). Vertical red lines show 95% confidence intervals. 

\pagebreak

```{r, fig.height = 3}

#custom function that calculates inter-baby differences over time (to avoid lengthy script)
interbaby_dist <- function(x, meth = 'bray') {
  
  #calculate distance matrix; rename rows and col to match sampleIDs
  ids <- x$sampleID
  sample_ts <- otus %>% 
    mutate(t = ceiling(t/30)) %>% 
    distinct(sampleID, t) %>%
    mutate(sampleID = factor(sampleID))
  x <- x[, -c(1)]
  x <- vegdist(x, method = meth)
  x <- as.matrix(x)
  row.names(x) <- ids
  colnames(x) <- ids
  
  #melt distances, remove distances from mirrored ids and self-id pairs,join t, remove the weak sampling after 36 months
  x <- melt(x, value.name = 'dist') %>%
    filter(as.numeric(factor(Var1)) > as.numeric(factor(Var2))) %>%
    left_join(sample_ts, by = c('Var1' = 'sampleID')) %>%
    rename(id1_t = t) %>%
    left_join(sample_ts, by = c('Var2' = 'sampleID')) %>%
    rename(id1 = Var1, id2 = Var2, id2_t = t) %>%
    filter(id1_t == id2_t & id1_t < 37)

  return(x)
}

# filter the rarest species out, group by genus-level abundances to speed up distance matrix calculation (ensure each genus has a unique code), create wide table
if (FALSE) {
  
  j1 <- otus %>%
    select(-subject, -t) %>%
    left_join(tax, by = 'otu') %>%
    group_by(sampleID, otu) %>%
    summarise(abun = sum(abun)) %>%
    spread(otu, abun, fill = 0) %>%
    interbaby_dist(meth = 'bray')
  
  saveRDS(j1, 'data\\otu_level_braycurtis.RDS')
  
} 

#I wonder if i can use the meltdist function from above to streamline this....
j1 <- readRDS(paste0(wd,'data\\otu_level_braycurtis.RDS'))

j2 <- otus %>%
    select(-subject, -t) %>%
    left_join(tax, by = 'otu') %>%
    filter(Order != 'unclassified') %>%
    group_by(Genus) %>%
    mutate(genusID = paste(Genus, as.numeric(as.factor(paste(Order, Family))))) %>%
    group_by(sampleID, genusID) %>%
    summarise(abun = sum(abun)) %>%
    spread(genusID, abun, fill = 0) %>%
    interbaby_dist(meth = 'bray')

#### Now the same thing for traits

j3 <- otus %>%
  mutate(otu = factor(otu)) %>%
  left_join(traits, by = 'otu') %>%
  gather(trait, val, -sampleID, -subject, -t, -otu, -abun) %>%
  group_by(sampleID, subject, t, trait) %>%
  summarise(val = weighted.mean(val, w = abun, na.rm = TRUE)) %>%
  group_by(trait) %>%
  mutate(val = scale(val)) %>%
  ungroup() %>%
  spread(trait, val, fill = 0) %>%
  select(-subject, -t) %>%
  interbaby_dist(meth = 'euclidean')

j <- rbind(mutate(j1, group = 'Taxonomic dissimilarity (OTUs)'), 
           mutate(j2, group = 'Taxonomic dissimilarity (genera)'),
           mutate(j3, group = 'Trait-based dissimilarity'))

j$group = factor(j$group, unique(j$group)[1:3])

j %>%
  rename(Months = id1_t) %>%
  filter(Months > 1 & Months < 36) %>%
  ggplot(aes(x = Months, y = dist)) +
    stat_summary(fun.y = mean, geom = "point", color = 'red') + 
    stat_summary(fun.data = mean_cl_boot, geom = "linerange", color = 'red') +
    stat_smooth(color = 'black', se = F, span = 10) +
    expand_limits(y = 0) +
    facet_wrap(~group, scales = 'free') +
    labs(x = 'Months after birth', y = 'Dissimilarity among infants') +
    th


```

## Figure 5. Mean dissimilarity in gut microbiomes across infants over early development.
Mean community dissimilarity among samples across all infants, grouped into one-month slices (excluding pairs of samples from the same infant). Taxonomic dissimilarity is quantified using Bray-Curtis dissimilarity using either OTUs (left) or genera (center) as the units of comparison; trait-based dissimilarity is calculated using Euclidean distance of all traits examined in this study, scaled so each trait has the potential for equal contribution to net dissimilarity.

\pagebreak

```{r, fig.height = 6}

otus_cs %>%
  left_join(traits, by = 'otu') %>%
  gather(trait, val, -sampleID, -subject, -t, -otu, -abun, -delivery) %>%
  group_by(sampleID, subject, t, trait) %>%
  summarise(val = weighted.mean(val, w = abun, na.rm = TRUE)) %>%
  left_join(meta, by = 'subject') %>%
  filter(trait %in% c('Obligate_anaerobe','Motility','Copies_16S','Forms_biofilms')) %>%
  ungroup() %>%
  mutate(t = round(t/30), t = ifelse(delivery == 'caesaren', t + 0.2, t)) %>%
  filter(t > 1.2 & t < 37) %>%
  mutate(delivery = ifelse(delivery == 'caesaren', 'Caesaren delivery','Vaginal delivery')) %>%
  ggplot(aes(x = t, y = val, color = delivery)) +
    stat_summary(fun.y = mean, geom = "point") + 
    stat_summary(fun.data = mean_cl_boot, geom = "linerange") +
    stat_smooth(se = FALSE) +
    facet_wrap(~trait, scale = 'free_y') +
    scale_color_discrete(name = '') +
    labs(x = "Months after birth", y = "Abundance-weighted community mean") +
    theme_bw() +
    theme(
      panel.grid.minor = element_blank(),
      panel.grid.major = element_blank(),
      panel.background = element_blank(),
      legend.position = 'bottom')


```

## Figure 6. Abundance-weighted trait patterns over time, grouped by delivery mode.
Circles show the average abundance-weighted mean of samples in each month (N = number of samples, ranging from and ranges from three - eight samples from a total of five infants born by Caesaren, or 26 - 47 samples from a total of 51 infants born vaginally. Vertical red lines show 95% confidence intervals. Colored trendlines illustrate changes in community weighted trait means over time, calculated using generalized additive models. For comparisons of abundance-weighted patterns by delivery mode for all traits, see Fig. S9.

\pagebreak

\begin{center}
\bigskip
\bigskip
\bigskip
\bigskip
{\Huge Supplementary tables and figures}
\end{center}

\pagebreak

\pagebreak

```{r trait inference, fig.height = 5}

tmp <- traits_sparse %>%
  gather(trait, val, -otu) %>%
  filter(!is.na(val)) %>%
  left_join(readRDS('data\\ggtax.RDS'), by = 'otu') %>%
  filter(Species != 'unclassified')

tmp_abuns <- otus %>% group_by(otu) %>% summarise(abun = sum(abun))  

#I drop traits that are already dependent on inferred trait data, from Bugbase
j <- traits %>%
  select(-Forms_biofilms, -Obligate_anaerobe, -Gram_positive) %>%
  filter(otu %in% otus$otu) %>%
  gather(trait, val, -otu) %>%
  distinct(otu, trait) %>%
  inner_join(tax, by = 'otu') %>%
  mutate(
    Species = paste(trait, Family, Genus, Species) %in% paste(tmp$trait, tmp$Family, tmp$Genus, tmp$Species),
    Genus = paste(trait, Family, Genus) %in% paste(tmp$trait, tmp$Family, tmp$Genus),
    Family = paste(trait, Family) %in% paste(tmp$trait, tmp$Family),
    Order = paste(trait, Order) %in% paste(tmp$trait, tmp$Order),
    Class = paste(trait, Class) %in% paste(tmp$trait, tmp$Class)
  ) %>%
  rename(C = Class, O = Order, F = Family, G = Genus, S = Species) %>%
  gather(level, coverage, C, O, F, G, S) %>%
  inner_join(tmp_abuns, by = 'otu') %>%
  mutate(level = factor(level, levels = rev(c('C','O','F','G','S')))) %>%
  group_by(trait, level) %>%
  summarise(coverage = sum(abun[coverage]) / sum(abun))
  
ggplot(j, aes(x = level, y = coverage)) +
  geom_bar(stat = 'identity') +
  facet_wrap(~trait) +
  theme_bw() +
  theme(
    panel.grid.minor = element_blank(),
    panel.grid.major = element_blank(),
    panel.background = element_blank(),
    legend.position = 'none') +
  labs(x = 'Maximum level of phylogenetic distance allowed when inferring trait data', y = 'Proportion of community with available trait data')

```

##Figure S1. Phylogenetic distances used when inferring traits
Unknown trait data were inferred by averaging trait values over descending subtrees in the greengenes phylogeny. That is, for each unknown trait value, the predicted value is set to the arithmetic average state of all tips descending from that node. On the x-axis, S = Species, G = Genus, F = Family, O = Order, C = Class. Missing trait data for three traits (biofilm forming, obgligate anaerobe status, and Gram positive status) were inferred in Ward et al. 2017, and are therefore not examined in this figure.

```{r pairwise trait correlations}

#I don't *want* to use the trait database with all the data extrapolated from congeners, but if I don't use it then there is *very* little to work with (which is why I did the extrapolation in the first place). And... in a sense... if we want to know which observations aren't independent, we *want* to use the extrapolated trait data. So that's what I'll do for the meantime. Really what we need are more trait data........

j <- cor(traits[, -c(1)], use = 'pairwise.complete.obs')
#j <- cor(traits_sparse[, -c(1:2)], use = 'pairwise.complete.obs')
j[lower.tri(j, diag = T)] <- NA

#two ways to look at correlations
melt(j) %>%
  filter(!is.na(value)) %>%
  ggplot(aes(x = Var1, y = Var2, fill = value)) +
    geom_tile() +
    scale_fill_gradient2(low = 'blue', mid = 'white', high = 'red', name = 'Pears. Corr.') +
    labs(x = '', y = '') +
    theme_bw() +
    theme(
      axis.text.x = element_text(angle = 45, hjust = 1),
      panel.grid.minor = element_blank(),
      panel.grid.major = element_blank(),
      panel.background = element_blank())

```

## Figure S2. Pearson correlation coefficients among traits.

\pagebreak

```{r sandbox, fig.height = 4}

otus_cs %>% 
  distinct(subject, t, delivery) %>% 
  left_join(meta, by = c("subject", "delivery")) %>% 
  arrange(desc(t_length)) %>%
  mutate(
    delivery = ifelse(delivery == 'caesaren', 'Caesaren delivery','Vaginal delivery'),
    Subject = as.numeric(factor(subject, levels = unique(subject)))
  ) %>%
  ggplot(aes(y = Subject, color = delivery)) + 
    geom_segment(aes(x = t_min/30, xend = t_max/30, yend = Subject)) + 
    geom_point(aes(x = t/30)) + 
    scale_color_manual(values = c('black','purple'), name = '') +
    th +
    labs(x = 'Months after birth', y = 'Subject')

```

##Figure S3. Sampling times, sampling durations, and delivery modes of infants used in this study.
Each horizontal line is a subject; each colored dot is a sample.

\pagebreak

```{r}
iwant <- c('Sporulation','Motility','Temp_optimum','GC_content')
otus %>%
  left_join(tax[, c('otu','Phylum')], by = "otu") %>%
  left_join(traits[, c('otu',iwant)], by = "otu") %>%
  mutate(Phylum = as.character(Phylum)) %>%
  gather(trait, val, -sampleID, -subject, -t, -otu, -abun, -Phylum) %>%
  group_by(sampleID, t, Phylum, trait) %>%
  summarise(val = weighted.mean(val, w = abun, na.rm = T)) %>%
  group_by(trait, Phylum) %>%
  filter(length(Phylum) > 10) %>%
  ggplot(aes(x = t/30, y = val, color = Phylum)) +
   stat_smooth() +
   labs(x = "Months after birth", y = "CWM sporulation score") +
   facet_wrap(~trait, scales = 'free') +
   th

```

## Figure S4. Changes in abundance-weighted mean trait scores over time, by Phylum.

\pagebreak

```{r delta plot, fig.height = 3.5}

j <- filter(j_delta, delta < 100)

#j %>% 
#  group_by(group) %>%
#  summarise(Pearson = cor(delta, dist))

ggplot(j, aes(x = delta, y = dist)) +
  geom_jitter(alpha = 0.1) +
  stat_smooth() +
  facet_wrap(~group) +
  th +
  labs(x = 'Dissimilarity', y = 'Days between samples')
  

```

## Figure S5. Dissimilarity in community composition plotted by the number of days between samples.
The Pearson correlation between OTU-based dissimilarity and the number of days between samples: r = 0.11 (N = 1118). The Pearson correlation between trait-based dissimilarity and the number of days between samples: r = 0.013 (N = 1118). Samples with more than 100 days between them (N = 30) were excluded.

\pagebreak

```{r all traits c section comparison, fig.height = 6.5}

otus_cs %>%
  left_join(traits, by = 'otu') %>%
  gather(trait, val, -sampleID, -subject, -t, -otu, -abun, -delivery) %>%
  group_by(subject, t, trait) %>%
  summarise(val = weighted.mean(val, w = abun, na.rm = TRUE)) %>%
  left_join(meta, by = 'subject') %>%
  filter(!trait %in% c('Obligate_anaerobe','Motility','Copies_16S','Forms_biofilms')) %>%
  ungroup() %>%
  mutate(t = round(t/30), t = ifelse(delivery == 'caesaren', t + 0.2, t)) %>%
  filter(t < 37) %>%
  mutate(delivery = ifelse(delivery == 'caesaren', 'Caesaren delivery','Vaginal delivery')) %>%
  ggplot(aes(x = t, y = val, color = delivery)) +
    stat_summary(fun.y = mean, geom = "point", alpha = 0.5) + 
    stat_summary(fun.data = mean_cl_boot, geom = "linerange") +
    stat_smooth(se = FALSE) +
    facet_wrap(~trait, scale = 'free_y', ncol = 2) +
    scale_color_discrete(name = '') +
    labs(x = "Months after birth", y = "Abundance-weighted community mean") +
    theme_bw() +
    theme(
      panel.grid.minor = element_blank(),
      panel.grid.major = element_blank(),
      panel.background = element_blank(),
      legend.position = 'bottom')


```

## Figure S6. Trait-patterns by delivery mode
Each circle represents the CWM of one sample for one infant; red and black circles are from infants born vaginally (N = 54 infants) and by Caesaren section (N = 7 infants), respectively bla. Black lines show community weighted mean (CWM) trait values of gut microbiota over time across all infants, grouped by delivery mode. Taxa without trait information were necessarily omitted when calculating CWMs; see Table 1 for coverage statistics. For abundance-weighted patterns for all traits, split by delivery mode, see Fig. S9.

\pagebreak

```{r prevalence and abundance} 

j <- otus %>% 
  left_join(meta[, c('subject','study')], by = 'subject') %>%
  mutate(t = ifelse(t < 365, 'Early', 'Late')) %>%
  group_by(otu, t) %>%
  summarise(
    abun = sum(abun),
    prevalence = length(unique(subject))) %>%
  group_by(t) %>%
  mutate(
    abundance = round(abun * 12),
    abundance = as.numeric(factor(abundance, levels = sort(unique(abundance))))) %>%
  mutate(
    Importance = scale(abundance) + scale(prevalence),
    Importance = Importance - min(Importance))

j1 <- filter(j, t == 'Early')

p1 <- ggplot(j1, aes(x = abundance, y = prevalence)) +
  geom_jitter(aes(color = Importance, alpha = Importance^2)) +
  scale_color_gradient(high = 'red', low = 'white', name = 'Potential\nImportance') +
  guides(alpha = 'none') +
    theme_bw() +
    theme(
      panel.grid.minor = element_blank(),
      panel.grid.major = element_blank(),
      panel.background = element_blank(),
      legend.position = 'bottom') +
  labs(x = 'Abundance rank', y = 'Prevalence') +
  ggtitle('Early successional taxa')

j2 <- filter(j, t == 'Late')

p2 <- ggplot(j2, aes(x = abundance, y = prevalence)) +
  geom_jitter(aes(color = Importance, alpha = Importance^2)) +
  scale_color_gradient(high = 'blue', low = 'white', name = 'Potential\nImportance') +
  guides(alpha = 'none') +
  theme_bw() +
  theme(
    panel.grid.minor = element_blank(),
    panel.grid.major = element_blank(),
    panel.background = element_blank(),
    legend.position = 'bottom') +
  labs(x = 'Abundance rank', y = 'Prevalence') +
  ggtitle("Late successional taxa")

mygrobs <- lapply(list(p1, p2), ggplotGrob)

gt <- gtable(widths=unit(c(1,1), "null"),
             heights=unit(1, "null"))

gt <- gtable_add_grob(gt, mygrobs, 
                       l=c(1,2),
                       r=c(1,2),
                       t=c(1,1),
                       b=c(1,1))
grid::grid.newpage()
grid::grid.draw(gt)

```

## Figure S7. Taxa abundance by prevalence: a measure of importance.
Importance is calculated as the sum of prevalence (scaled) and abundance rank (scaled).

\pagebreak

```{r important taxa table 1}

tmp <- j1 %>%
  ungroup() %>%
  left_join(tax, by = 'otu') %>%
  select(Family, Genus, Species, OTU = otu, Importance) %>%
  mutate(Importance = round(as.numeric(Importance), 2)) %>%
  arrange(desc(Importance)) %>%
  filter(seq_along(Importance) < 21)

kable(tmp, format = 'latex', escape = F, booktabs = T, linesep = "") %>%
  kable_styling() %>%
  row_spec(0, bold = T) %>%
  footnote(general = "Early successional taxa, sorted by importance; importance is calculated as the sum of prevalence (scaled) and abundance rank (scaled).", general_title = "Table S1.", footnote_as_chunk = T, threeparttable = TRUE)


```

\pagebreak

```{r important taxa 2}

tmp <- j2 %>% 
  ungroup() %>%
  left_join(tax, by = 'otu') %>%
  select(Family, Genus, Species, OTU = otu, Importance) %>%
  mutate(Importance = round(as.numeric(Importance), 2)) %>%
  arrange(desc(Importance)) %>%
  filter(seq_along(Importance) < 21)

kable(tmp, format = 'latex', escape = F, booktabs = T, linesep = "") %>%
  kable_styling() %>%
  row_spec(0, bold = T) %>%
  footnote(general = "Late successional taxa, sorted by importance; importance is calculated as the sum of prevalence (scaled) and abundance rank (scaled).", general_title = "Table S2.", footnote_as_chunk = T, threeparttable = TRUE)

```

