---
bibliography: C:\\Users\\John\\Documents\\library.bib
csl: C:\\Users\\John\\Documents\\ecology.csl
output:
  bookdown::word_document2:
    reference_docx: C:\\Users\\John\\Documents\\rmd_style_template.docx
    toc: no
  word_document:
    toc: no
  bookdown::html_document2:
    fig_caption: yes
    theme: united
    toc: yes
    toc_float:
      collapsed: no
      smooth_scroll: yes
---

```{r initialize manuscript, include=FALSE}

#set knitr options
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE, cache = FALSE)

library(scales)

#initialize manuscript formatting script
source('C:\\Users\\John\\Documents\\initialize_manuscript.R')

#load data, create figures
source('C:\\Users\\John\\Documents\\msu\\microbiome_trait_succession\\succession_analysis.R')

#set directory where figures will be loaded/generated
#note that kable_as_image doesn't work if there are underscores in the headers...
image_dir <- 'C:\\Users\\John\\Documents\\msu\\microbiome_trait_succession\\images\\'

```

#Trait-based community assembly and succession of the infant gut microbiome

\newline

John Guittar^1\*^, Ashley Shade^2,3^, Elena Litchman^1,3,4^

\newline

^1^ Kellogg Biological Station, Michigan State University, 3700 E Gull Lake Dr., Hickory Corners, MI 49060, USA

^2^ Department of Microbiology and Molecular Genetics, Department of Plant, Soil and Microbial Sciences, and The Plant Resilience Institute, Michigan State University, East Lansing MI 48840, USA

^3^ Program in Ecology, Evolutionary Biology and Behavior, Michigan State University East Lansing MI 48840, USA

^4^ Department of Integrative Biology, Michigan State University, East Lansing, MI 48824, USA

\newline

^\*^ Correspondence author: E-mail: guittarj@msu.edu

\newline

*Keywords*: Community assembly, dispersal limitation, ecological succession, functional ecology, gut microbiome, human microbiome, microbial ecology, trait-based ecology.

\newline

#Abstract

The microbial community in the human gut develops over the first three years of life and plays important roles in modulating human digestion and overall health. Yet our understanding of the mechanisms driving gut community succession is limited. Trait-based approaches have proven useful for studying community assembly in macroorganisms by linking community structure and function, but have not yet been systematically applied to gut microbiomes. Here we examine phenotypic and genotypic trait patterns over gut succession in `r nrow(meta)` human infants. We compile Bacterial trait data from the literature and several online databases, and use hidden state character prediction to estimate unknown trait values. Mirroring other microbial successional systems, mean 16S rRNA gene copy number decreased over time, implying a reduction in the ecological importance of rapid growth. Sporulation ability, a trait that putatively promotes among-host dispersal, increased over time, whereas motility, a trait that can promote within-host colonization, decreased over time. Bacterial environmental preferences also evidently shifted over infant development, including a decrease in oxygen tolerance and an increase in optimal growth temperature. Trait-based community composition effectively stabilized after the first year, while taxonomic turnover proceeded throughout infant development, indicating that community turnover after the first year is functionally redundant with respect to the traits examined in this study. Both taxonomic and, especially, trait-based community composition converged across infants over time, consistent with an assembly process in which a stochastic set of initial taxa is gradually replaced by taxa that are better suited for the gut environment.

\newline

#Introduction

Classical ecological theory posits that successional patterns arise from the combined influence of dispersal, species interactions, and the environment [@Pickett1987b; @Walker1987a; @Tilman1990a; @Hubbell2001a; @Louca2018], and this general framework extends readily to gut communities [@Lozupone2012b]. Before a microbe can inhabit the colon, the most distal and speciose part of the gastrointestinal tract, it must first be swallowed by the host and survive the acidic conditions of the stomach and small intestine (i.e., it must disperse). A species will persist in the colon only if it can acquire enough resources to reproduce (i.e., it must be competitive) or arrive there in high enough numbers to sustain a population [@Leibold2004]. Microbial colonists may then alter the environment, e.g., by depleting intestinal oxygen [@Backhed2015] or providing opportunities for cross-feeding [@Belenguer2006], favoring taxa with different phenotypes as succession proceeds.

`r indent` Yet, the process of gut succession may conform poorly to classical successional patterns [@Connell1977b; @Tilman1990a; @Fierer2010b] due to the active influence of the host and the host mother [@Adair2017; @Foster2017]. Early colonists are passed directly from the mother during or even before birth [@Moles2013b; @Stout2013a; @Ardissone2014a], and therefore may lack characteristics that would otherwise facilitate early arrival, e.g., via active dispersal, and instead have characteristics selected for in the mother's gut or vaginal environment. Following birth, mothers supply bacterial growth factors in breastmilk and continue to introduce new taxa through physical contact [@Fernandez2013c]. Meanwhile, the maturing infant begins to suppress undesirable taxa through immune response, and actively cultivate commensal taxa by providing a steady supply of nitrogen-rich mucus and potential habitat in the outer mucus layer of the large intestine [@Li2015a]. Gut community composition is also affected by the introduction of solid food [@David2013a; @Walter2015a], in particular with the introduction of insoluble fiber [@Valles2014a].  

`r indent` Gut community successional patterns will necessarily reflect a combination of dispersal, microbial species interactions, and host physiology and behavior. A present challenge is to determine the relative importance of these drivers and how their importance changes over time. One approach to disentangling the mechanisms of community assembly is to examine patterns in trait-based community composition [@McGill2006]. A trait, in the broadest sense, is defined as a measurable organismal characteristic directly or indirectly linked to fitness or performance [@Violle2007; @Green2008b; @Litchman2015; @Martiny2015]. Observable shifts in the trait-based composition of a community imply shifts in local environmental conditions favoring different species and/or dispersal limitation (i.e., when a taxon does not colonize a site because it does not arrive). Despite the success and proliferation of trait-based approaches in studying community assembly in plant, animal, and some microbial systems [@Suding2003; @Wright2004; @McGill2006; @Violle2007; @Litchman2008; @Guittar2016a], they are less often used to study bacterial and archaeal systems [but see: @Green2008b; @Allison2012a; @Ortiz-Alvarez2018]. This is partly due to the challenges of identifying ecologically relevant traits for a functionally diverse cohort of taxa, and partly due to a dearth of curated trait data. However, thanks to advances in high-throughput molecular techniques, renewed efforts to directly collect phenotypic data [@Browne2016], and the aggregation of data from disparate sources [@Sohngen2016; @Barberan2017], trait-based approaches to microbial community dynamics are becoming more feasible, especially in well-studied systems like the human gut.

`r indent` Here, we examined trait-based successional patterns in a cohort of `r nrow(meta)` infants from Finland and Estonia [@Kostic2016a; @Yassour2016d] and type-1 diabetes using an expanded set of curated microbial traits. To populate our trait database, we (1) built a phylogeny comprising the taxa from the gut communities and 13900 type-specimens with known Latin binomials [@Munoz2011], (2) mapped trait data onto the tips of the phylogeny by matching Latin binomials, and (3) used hidden state prediction to infer unknown trait values. We only used inferred trait values when the distance of phylogenetic inference fell below the threshold at which phylogenetic signal could be detected for each trait. This unique approach maximized the use of existing trait data while minimizing unwarranted predictions. We then compare taxonomic and trait-based community turnover in space and time to shed additional light on the mechanisms driving successional patterns. The examined microbial traits relate to maximum growth rate and resource use strategy (e.g., 16S rRNA gene copy number), environmental preferences (e.g., optimal temperature for growth), dispersal ability (e.g., sporulation), stress tolerance (e.g., oxygen tolerance), and host benefit (e.g., number of B vitamin synthesis pathways in the genome), with additional traits described below. We ask three sets of questions related to the mechanisms underlying gut community assembly and succession:

1. Are there consistent trait-based patterns of succession? If so, what do they suggest about how successful microbial strategies and/or host influences change over time? How do differences in initial conditions, e.g., due to delivery by cesarean section (C-section), and disturbance regimes e.g., antibiotic treatments, affect trait-based successional patterns?

2.	How closely linked are taxonomic and trait-based successional patterns? To what extent do taxonomic changes imply trait-based changes?

3.	To what degree is gut community succession functionally and taxonomically predictable? Do infant gut communities converge in traits and taxa over time, as expected in a generally deterministic system, or fail to converge, as expected in a more stochastic community assembly process contingent on which species arrive and/or divergent host environments?

\newline

#Methods

##Infant microbiome sampling and sequence processing

Our foremost aim in this study was to characterize general patterns of gut primary succession that hold true regardless of host-related differences. As such, unless otherwise noted, we include all infants in our analyses, regardless of delivery mode or other host differences specific to each included study. Longitudinal infant gut microbiome data were compiled from two studies from the DIABIMMUNE study group [(https://pubs.broadinstitute.org/diabimmune](https://pubs.broadinstitute.org/diabimmune), one focused on the effects of antibiotics on gut community development [@Yassour2016d], and the other focused on the effects of type-1 diabetes on gut community developmnet [@Kostic2016a]. Infants from @Yassour2016d either had nine or more antibiotic on gut community development courses, or no antibiotic courses. Infants from @Kostic2016a tested positive for HLA DR-DQ alleles conferring risk of type-1 diabetes; of the infants which met our sampling criteria (see below), three developed type-1 diabetes during the sampling period.

`r indent` Stool samples of infants were collected by participants' parents and stored in the house freezer until the next scheduled visit to the local study center. Samples were then shipped on dry ice to the DIABIMMUNE Core Laboratory, where they were stored at -80°C until being sent to the Broad Institute for DNA extraction and 16S rRNA amplicon sequencing. Sequencing was performed on the Illumina HiSeq 2500 platform using the 515F and 806R primers. Of 74 infants across the two studies, only those with at least 12 samples and those which extended more than 30 months were used in this study, yielding `r nrow(meta)` infants with `r sampstats$min` - `r sampstats$max` sampling points (mean = `r sampstats$mean`; median = `r sampstats$median`) taken at semi-regular intervals over the first 3 years of infant life ((ref:figSampling)). All subjects were from Finland, except one from Estonia. 

`r indent` Infants varied in their modes of delivery and antibiotic histories, providing an opportunity to explore the potential effects of these "treatments" on trait-based gut community composition. To this end, infants were divided into three groups: 1) High antibiotic exposure (N = `r N_ab`), if they underwent at least 50 days of antibiotic treatment and were delivered vaginally, 2) C-section delivery (N = `r N_cs`), if they were delivered by C-section and underwent two or fewer rounds of antibiotics, and 3) a "Control" group (N = `r N_control`), if they were born vaginally and received no antibiotic treatments. In some instances, antibiotic treatment durations were not reported, in which case we assumed seven days per treatment. Twelve types of antibiotics were administered for a variety of ailments, with the most common being amoxicillin, trimetoprime and sulfadiazine aimed at treating acute ear infections. For additional background and methodological details see @Kostic2016a and @Yassour2016d.

`r indent` Sequence processing was done using USEARCH v10.0.240 [@Edgar2010]. Sequencing data were downloaded from the DIABIMMUNE website [https://pubs.broadinstitute.org/diabimmune/](https://pubs.broadinstitute.org/diabimmune/). Chimeras and reads flagged with more than one error were excluded, and the remaining reads were truncated to 250 bp, the expected overlap when using 515F and 806R primers. Reads were clustered into operational taxonomic units (OTUs) at 97 percent sequence identity using the UPARSE-OTU algorithm. Representative sequences from each OTU were mapped to the SILVA v123 database [@Quast2013a] to determine potential taxonomic identities. To avoid bias in sampling effort, samples were rarefied to 5000 sequences, and seven samples with fewer than 5000 sequences were removed.

\newline

##Assembling trait data

We compiled data on 16 genomic, physiological, and life history traits of bacteria from public databases and individual studies ((ref:tabTraitSources), (ref:supdatTraitObs)). All trait data were associated with taxa with full Latin binomials (i.e., Genus and Species labels) that appeared either in the SILVA-derived taxonomy file for the combined gut community samples or in the curated taxonomy file from the 132 release of the Living Tree Project [@Munoz2011]. Altogether, these amounted to `r comma(sparse_tdats)` collected trait data spread across `r comma(nrow(traits_sparse))` taxa. When a taxon had more than one trait value, the mean or mode was used, depending on whether the trait was quantified continuously or discretely.

`r indent` Descriptions and data sources for each trait are listed briefly in (ref:tabTraitSources), but here we elaborate with a few additional details: 1) The numbers of B-vitamin synthesis pathways in the genome were determined by @Magnusdottir2015 and were based on genome annotations from the pubSEED platform [@Overbeek2014]. 2) The number of genes in the genome correlated strongly with genome size (Pearson's $r =$ `r gene_genome_corr`), so genome size was dropped from subsequent analyses to avoid redundancy. 3) In some cases, optimal temperature was calculated as the mean of lower and upper temperature ranges (consistent with @Barberan2017). 4) IgA binding affinity refers to the degree that immunoglobulin A (IgA) bound to specific bacterial taxa, and was calculated as an "IgA coating index" by @Palm2014 using flow-cytometry-based bacterial cell sorting and 16S rRNA sequencing to characterize the coating load of IgA on specific taxa from fecal samples in a murine model. 5) Sporulation score indicates the tendency of taxa to sporulate, and was calculated by @Browne2016 as a continuous score ranging from zero that depended on a combination of targeted phenotypic culturing and whole-genome sequencing from stool samples.

\newline

##Predicting unknown trait data

We estimated unknown phenotypes and genotypes using hidden state prediction methods based on phylogenetic inference. Specifically, we first generated a phylogenetic tree with the 3,311 OTUs from our USEARCH pipeline (pre-rarefying) and the 13,900 OTUs from the 132 release of the Living Tree Project (LTP) [@Munoz2011] ((ref:figTreeTraits)). The topology of the tree reflects percent sequence similarity among taxa in the 16S rRNA V4 region, and was generated using agglomerative clustering of a distance matrix based on the U-sort heuristic [@Louca2017]. Because the LTP representative sequences comprised the entire 16S rRNA gene (ribosomal small subunit), they first needed to be truncated to the 250 bp of the V4 region using 515F and 806R primers before generating the distance matrix. Trait data were then mapped onto the tips of the resultant phylogenetic tree using Latin binomials. The LTP database is uniquely well-suited to interface with literature-derived trait data because each sequence represents a type strain with Genus and Species annotations drawn from the literature, not inferred phylogenetically.

`r indent` Missing trait values were estimated using three algorithms for hidden state prediction: independent contrasts, subtree averaging, and weighted squared-change parsimony, using the R package 'castor' [@Louca2017]. The three methods have different strengths and weaknesses [@Zaneveld2014; @Louca2017], but their predictions correlated strongly ((ref:tabHSP)), lending confidence to the results. We opted for weighted square-change parsimony for our analysis, which recursively calculates locally parsimonious states for each node based on its descending subtree, until reaching a parsimonious state estimate for the tree root [@Maddison1991]. Because we treated all trait values as numeric (e.g., Gram-negative = 0 and Gram-positive = 1), state predictions for discrete traits could still be fractional (e.g., a Gram-positive score of 0.5), reflecting their probabilistic uncertainty. The trait value predictions used in this study are available in (ref:supdatTraitPreds).

`r indent` A risk of using hidden state prediction is that state estimates are generated for all taxa with hidden states, even when there is insufficient confidence to warrant prediction. To mitigate this, we examined how trait dissimilarity varied with increasing phylogenetic distance, and only used predictions when there were closely-related taxa with known trait values (see (ref:figThresholds) for a graphical aid to understanding the approach). Specifically, for each trait, phylogenetic trees were generated using all OTUs with known trait data. Next, differences in trait values and phylogenetic distance (i.e., percent 16S rRNA V4 sequence similarity) were calculated for all possible OTU pairs. In some cases, trait-specific trees were prohibitively large, and only 10,000 OTU pairs were randomly selected at each 0.005 increment phylogenetic distance. Five generic models were then used to predict trait differences, $|y|$, as a fuction of phylogenetic distance, $x$, and the best fitting model was selected by AIC. The models included: 1) Null: $|y| \sim 1$; 2) Linear regression: $|y| \sim x$; 3) Asymptotic regression: $|y| \sim a(1 - e^{(-e^bx)})$, where $a$ and $b$ were determined using a self-starting nonlinear least squares approach, and the model fit was constrained to pass through the origin; and 4) Logistic regression: $|y| \sim \frac {a}{1 + e^{(\frac{b-x}{c})}}$, where $a$, $b$, and $c$ were determined using a self-starting nonlinear least squares approach. Null models provided the best fit for Aggregation score, Salt optimum, and IgA binding affinity, indicating that for these traits it was not possible to estimate missing trait data at any phylogenetic distance. An asymptotic regression model provided the best fit for pH optimum, but the model was qualitatively equivalent to the null model, so predictions for this trait also not considered. For the remaining 11 traits, trait predictions were only used when taxa with known trait values occurred within trait-specific thresholds of phylogenetic distance; we defined these thresholds as the points at which model predictions rose to 90 percent of null expectations ((ref:tabThresholds)). Null expectations equaled the mean trait-based differences of all OTU pairs with more than 0.1 phylogenetic distance between them, for each trait. Overall, for traits that were amenable to hidden state prediction, this approach yielded trait data for `r tc$coverage[1]` percent (`r tc$trait[1]`) to `r tc$coverage[2]` percent (`r tc$trait[2]`) of sequences in this study ((ref:figTraitCoverage)). We assessed statistical independence among traits using Pearson correlation coefficients; p-values for correlations were adjusted by multiplying by 55, the number of pairwise combinations of traits with sufficient data examined in this study.

\newline

##Characterizing trait-based successional patterns and differences among infant cohorts

Trait-based successional patterns were evaluated at both the OTU-level and the community-level. For the OTU-level analysis, OTUs were assigned to one of three successional stages using linear models of OTU abundances over time across all infants: "early successional" OTUs had statistically significant negative trends in abundance over time (p < 0.05, $\beta$ < 0); "late successional" OTUs had positive trends in abundance over time (p < 0.05, $\beta$ > 0); taxa that did not fall in to early or late successional stages were labeled as "mid successional / no trend," encompassing taxa with sporadic, unvarying, or hump-shaped patterns of abundance over time. Statistical differences between the trait values of groups were determined with Welch t-tests; p-values were adjusted by multiplying by 22, equal to the number of traits multiplied by the number of non-control treatments.

`r indent` Trait-based differences at the community level were quantified using community weighted means (CWMs). A CWM is the mean trait value of the OTUs in a community, weighted by their abundances. Formally, a CWM is equal to $\sum_{i=1}^S p_i x_i$, where $p_i$ is the abundance of OTU $i$ ($i=1,2,...S$), and $x_i$ is the trait value for species $i$. Ecologically speaking, CWMs define the dominant traits of a community, and can be thought of both in terms of how they reflect system properties (i.e., as response traits) and how they influence system properties (i.e., as effect traits)[@Lavorel2002].

\newline

##Comparison of taxonomic and trait-based dissimilarity and turnover

We quantified differences in microbiome community compositions in two ways. First, we used Bray-Curtis dissimilarity to quantify differences in the OTU-based compositions of samples [@Beals1984]. Second, we quantified trait-based differences among communities with multidimensional Euclidean distance [@Petchey2002]. Specifically, Euclidean distance between two communities was calculated by (1) scaling trait values by their standard deviations to give each trait equal weight, (2) calculating the CWMs of each trait for both communities, and then (3) using the Pythagorean theorem to determine the distance between the two communities in n-dimensional trait space.

`r indent` We examined OTU-based and trait-based community changes over time in two ways. First, we examined compositional differences between subsequent samples from the same infant, at intervals approximately between one to three months, to quantify variation in rates of community turnover over infant development. We termed this "instantaneous turnover." Second, we examined compositional differences between samples and the final sample from each infant to quantify rates of "directional turnover" over infant development. To determine how trait-based rates of instantaneous and directional turnover compared to those expected by chance (i.e., if community dynamics were not influenced by traits), we compared observed data to a null model of trait-agnostic community change. Specifically, we generated 1000 mock versions of our data with trait values randomly shuffled among OTUs, and repeated the calculations described above. We then tested for differences between observed and null turnover rates using Welch t-tests. To test if observed rates of directional turnover were different than those expected by chance, we compared the magnitudes of the decreases in trait-based distance between subsequent six-month periods for both observed and null data sets, and again used Welch t-tests to determine statistical significance.

`r indent` To determine if community composition converged or diverged across infants as they aged, we divided samples into one-month slices and calculated the mean OTU-based and trait-based distances for all pairwise combinations of samples from different infants (i.e., we excluded any comparisons of samples from the same infant). Similar to the analysis of trait-based rates of change within infants, we compared observed data to 1000 null model replicates of trait-agnostic community change (i.e., data for which trait values and OTUs were randomly shuffled) in order to determine how rates of trait-based convergence across infants compared to those expected by chance. We tested for statistical differences between observed and null data during each six-month period with Welch t-tests; p-values were adjusted by multiplying by 11, the number of traits considered.

\newline

## Data availability

Raw sequencing data are available online at the NCBI project accession numbers [PRJNA231909](https://www.ncbi.nlm.nih.gov/bioproject/?term=PRJNA231909) and [PRJNA290381](https://www.ncbi.nlm.nih.gov/bioproject/?term=PRJNA290381). Custom scripts used in the bioinformatic pipeline and statistical analyses are available at [https:/github.com/ShadeLab/microbiome_trait_succession](https:/github.com/ShadeLab/microbiome_trait_succession). All trait data gathered or generated in this study are included as supplementary data files. 

```{r, include = FALSE}

ref(type = 'supdat', name = 'supdatTraitObs')
ref(type = 'supdat', name = 'supdatTraitPreds')

```

(ref:supdatTraitObs) Supplementary Data `r refs$supdatTraitObs`
(ref:supdatTraitPreds) Supplementary Data `r refs$supdatTraitPreds`

\newline

#Results and Discussion

##Are there consistent trait-based patterns of succession?

We observed consistent taxonomic and trait-based shifts in infant gut microbiomes during the first three years of infant life ((ref:figTaxOverTime), (ref:figCWMs), (ref:figCWMsSuccGroup)). Early succession was dominated by Bacterioidaceae and Bifidobacteriaceae, whereas late succession was dominated by Lachnospiraceae, Ruminococcaceae, and (still) Bacteroidaceae ((ref:figTaxOverTime)). @Kostic2016a and @Yassour2016d provide more thorough analyses of successional taxonomic patterns. Broadly, the observation that ~ 75 percent of OTUs exhibited consistent trends in abundance over succession speaks to the non-random nature of gut microbiome development. In line with this, early and late successional specialists differed significantly in their traits: late successional specialists were less tolerant of oxygen, were more likely to sporulate, and had higher temperature optima than early successional specialists ((ref:figCWMsSuccGroup)). Trait CWMs also trended significantly over the course of succession ((ref:figCWMs)), with the most pronounced shifts being decreases in GC content, oxygen tolerance, and flagellar motility, and increases in sporulation and optimal temperatures for growth. The mean number of B-vitamin pathways among taxa also decreased over time, contradicting the hypothesis that human hosts would selectively favor such taxa over the course of succession to promote the production of these essential nutrients. The number of 16S rRNA gene copies decreased steadily over development, consistent with successional patterns observed in many other microbial systems, believed to reflect a decrease in the competitive advantage of maximum growth rate [@Ortiz-Alvarez2018].

`r indent` Many traits correlated significantly among taxa, even after performing a Bonferroni correction to account for the large number of comparisons ((ref:figTraitCorr)). The strongest positive correlations were between sporulation and Gram-positive status, oxygen tolerance and GC content, and gene number and the number of B-vitamins, while the strongest negative correlations were between optimal growth temperature and oxygen tolerance, Gram-positive status and B-vitamin pathway number, and GC content and the number of 16S rRNA gene copies. Most of the remaining trait pairs had Pearson correlation coefficients less than 0.5 or greater than -0.5. On one hand, strong correlations among traits are noteworthy because they may reflect traits that covary because they influence a given taxon's position on the same ecological tradeoff axis (i.e., trait syndromes). For example, the negative correlation observed between sporulation and oxygen tolerance are two approaches to dealing with oxidative stress, either by becoming dormant until oxidative stress is relaxed, or by tolerating it. On the other hand, correlations among traits could simply be an artifact of arbitrary trait linkage, and not evidence for a given mechanism. As such, the mechanisms we discuss in this study are hypotheses, not conclusions, and we hope they spur further experimental work.

`r indent` The observed shifts in trait-based community composition over succession ((ref:figCWMs)) strongly suggest that non-random ecological processes are at work. Notably, shifts in two traits potentially related to dispersal ability suggest that dispersal dynamics may play a key role in shaping successional patterns. First, the initial presence and subsequent decline of taxa with flagella could mean that the ability to actively disperse over short distances (e.g., within hosts) improves colonization rate during early succession, but that flagella are not necessarily advantageous in the mature gut. In support of this, experiments have shown unflagellated strains to be poorer colonizers of chickens' gastrointestinal tracts than flagellated strains [@Nachamkin1993], and a positive relationship between motility and bacterial transmission [@Josenhans2002].

`r indent` Second, an increase in sporulating taxa over time ((ref:figCWMs), (ref:figSporulatorsTime)) may reflect the long-term advantages of being able to disperse among hosts and/or persist within hosts in a dormant state during stressful conditions [@Lennon2011; @Browne2016]. As succession proceeds and the gut environment becomes increasingly anaerobic, obligate anaerobes gain a competitive advantage over facultative anaerobes because they do not need to maintain the machinery for tolerating oxidative stress. However, this advantage presumably comes at the cost of being more vulnerable to oxidative stress while dispersing through aerobic environments to colonize new hosts. Sporulating taxa circumvent this putative tradeoff by traversing aerobic environments as oxygen-tolerant spores, and then thriving in the gut as obligate anaerobes. Hence, the observed increase of sporulating taxa over gut community development, both in total abundance ((ref:figCWMs)) and OTU richness ((ref:figSporulatorsTime)), may reflect the steady arrival and successful colonization of these well-adapted gut taxa. 

`r indent` Community-weighted mean 16S rRNA gene copy number ((ref:figCWMs)), a trait associated with the ability to quickly exploit available resources, possibly due to higher maximum growth rates [@Klappenbach2000b], decreased steadily over time. A decrease in 16S rRNA gene copy numbers over time is characteristic of primary succession in many microbial systems that are initially rich in resources [@Fierer2010b; @Kearns2018], such as a vial of sterile nutrient broth placed in an open-air environment [@Nemergut2016]. This pattern could arise if faster growing taxa thrive during early succession, when the primary carbon source is easily-digested milk or formula, and slower growing taxa only begin to thrive in mid to late succession, as the primary carbon source shifts towards increasingly complex molecules derived from solid food. Such a tradeoff could reflect a more general tradeoff between rapid growth and resource use efficiency [@Litchman2015a] which may be an important driver of ecological succession in other systems [@@Tilman1990a].

`r indent` To explore how early exposure to different taxa could affect gut succession, we compared trait-based successional patterns of infants delivered vaginally and by C-section ((ref:figCWMsTreat)). We reasoned that any consistent community differences between treatments likely arose due to differences in early colonization, i.e., because infants born vaginally were initially colonized by taxa from the mother during delivery, and infants born by C-section were initially colonized by taxa predominantly from the ambient environment (e.g., the mother's skin, hospital surfaces). Of several significant trait-based differences between C-section and vaginally born infants, the most notable were initially elevated mean 16S rRNA gene copy numbers, initially greater oxygen tolerance, and initially more flagellated taxa, and these differences generally persisted for about 20 months. At minimum, these results suggest that taxa encountered during vaginal delivery are functionally distinct from taxa from the surrounding ambient environment. More interestingly, however, they indicate that gut colonization patterns arise from an interaction between ecological filtering and the composition of the "regional" pool of available species. Systematic trait-based differences by birth mode persisted for up to two years, and perhaps longer in the case of oxygen tolerance, indicating that differences in early colonization can have lasting effects on community composition, a phenomenon also termed priority effects [@Sprockett2018; @Fukami2011b], consistent with other studies [@Salminen2004; @Azad2013c; @Dogra2015b]. The sustained differences between infants by delivery mode are somewhat surprising given recent work by @Ferretti2018, which found strong selective forces to quickly discourage immigrant taxa from the mother's skin or birth canal; such findings suggest that the persistent differences may be due to a lack of arrival (i.e., dispersal limitation) of gut-adapted taxa, rather qualitatively different community filters among infants.

`r indent` Exposure to antibiotics correlated with consistent trait-based shifts, especially during late succession ((ref:figCWMsTreat)). Most notably, taxa in infants exposed to antibiotics were, by the third year of development, less likely to be Gram-positive and less likely to sporulate. The drop in the relative abundances of Gram-positive taxa over time is expected, given that Gram-positive taxa lack the protective outer membrane that make Gram-negative bacteria generally more resistant to antibiotics [@Mcdonnell1999]. The drop in mean sporulation score was less expected, because spores are generally very resistant to antibiotics [@Mcdonnell1999]. However, spore formation is far from the only mechanism of antibiotic tolerance in Bacteria, and other strategies may be more effective for survival in the gut environment. For example, antibiotics are known to negatively affect the class Clostridia, a spore-forming taxonomic group of Firmicutes, while favoring Enterobacteraceae, a non-spore forming family [@Rivera-Chavez2016]. More generally, consistent with prior work [@Jernberg2007;@Jernberg2010], the persistent differences in trait-based community composition between infants that underwent heavy antibiotic treatments and those that did not, suggests that these disturbances can exert long-term effects on community structure and function.

`r indent` Trait variances within infant gut communities decreased in seven traits (e.g., in 16S rRNA gene copy number, GC content, gene number, length, motility, oxygen tolerance, and temperature optimum and motility) and increased notably in only two traits (sporulation score and width) ((ref:figCWMsVariance)). The overall decrease in trait-based variance over time indicates that individuals of the gut community became more functionally homogeneous as the infants matured, perhaps due to increasingly strict environmental filtering processes [@Cavender-Bares2004; @Cornwell2009] and/or competitive exclusion of poorly adapted taxa [@Mayfield2010].

\newline

##How closely tied are taxonomic and functional changes over succession?

To evaluate the degree to which taxonomic changes could serve as proxies for trait-based changes, we compared taxonomic and trait-based turnover over time within infants, both in terms of instantaneous turnover (i.e., differences between subsequent samples) and directional turnover (i.e., distance to the final, most "mature" sample for each infant). Rates of instantaneous turnover were higher over early succession in both OTU-based and trait-based measures of community dissimilarity, until falling to "baseline" levels at about 18 months for the rest of the sampling period ((ref:figIntrababyDiss)). A drop in species turnover over time is a feature of ecological succession in many other systems [@Anderson2007]. Although trait-based rates of instantaneous turnover exceeded null model expectations, these differences were neither large nor significant, suggesting that trait-based differences between subsequent samples are not much greater than those expected by chance, given the observed rates of instantaneous turnover in OTUs.

`r indent` Infant gut communities matured much faster in terms of their traits than their OTU-based compositions ((ref:figIntrababyDiss)). Specifically, directional turnover in OTUs was relatively steady across all three years of study, whereas directional turnover in traits was high only in the first year before dropping to nearly-baseline levels of instantaneous turnover. We compared the observed rates of trait-based directional turnover to a null model to determine statistical support for this data-based hypothesis that trait-based directional turnover exceeded that expected by chance, given the observed rates of OTU-based directional turnover. These results suggest that the infant gut microbiome becomes more stable (low turnover) in traits and their underlying functions before it becomes stable in OTU-based taxonomic composition, aligning with previous metagenomic work [@Koenig2011]. The steady OTU-based directional turnover over infant development, despite convergence in trait-based community composition after the first year, suggests that turnover in late succession is of taxa that are functionally redundant with respect to the traits examined in this study. Functionally redundant turnover could arise through a combination of variable immigration rates and ecological drift (i.e., changes in the relative abundances of taxa through stochastic birth/death events). Although the gut community has a large number of individuals, which, all else equal, makes it less susceptible to drift [@Louca2018], many of its taxa are rare and therefore still vulnerable to stochastic variation in their relative population sizes over space and time. Future work should work to quantify and characterize immigration rates, and consider other traits as potential drivers of late-stage successional community turnover, such as those relating to metabolism of specific dietary compounds, cross-feeding, or phage-host interactions [@Belenguer2006; @Corfe2015c; @Reyes2013].

\newline

##Is gut community succession predictable across individuals?

Surprisingly, gut community compositions became more similar (i.e., converged) across infants as they aged, both in terms of OTUs and traits ((ref:figInterbabyDiss)). This ran counter to our expectations that communities would diverge as infants shifted from subsisting on milk and/or formula (low variability expected among hosts), to easily digested soft foods (some resource variability expected among hosts), to solid foods (most resource variability expected among hosts), and as infants began interacting with idiosyncratic family-specific local environments. Instead, community convergence over time may reflect a process whereby a stochastic set of initial taxa are gradually replaced, or supplemented with, taxa that are better suited for the gut environment. Alternatively, convergence across infants could occur if mothers harbored, and thus imparted to their children at birth, microbial taxa that were compositionally more dissimilar than taxa that colonize infant guts during early childhood. In either case, it seems likely that community convergence over time is at least partly due to delayed arrival of taxa best adapted for the gut environment. Future experimental work should quantify the relative importance of dispersal dynamics and niche availability in driving compositional convergence over time.

`r indent` Compositional convergence among infant gut communities was more pronounced and abrupt in trait-based composition than in OTU-based composition, which converged only slightly and gradually over time ((ref:figInterbabyDiss)). To statistically test this data-based hypothesis, we compared observed rates of trait-based convergence to a null model in which trait values were randomly shuffled among OTUs ((ref:figInterbabyDiss)). The discrepancy between OTU-based and trait-based convergence leads to two insights. First, it again suggests a disconnect between community structure and functional potential, similar to patterns of directional turnover over time ((ref:figIntrababyDiss)). Second, it means that community composition was more predictable with respect to traits than OTUs, and that OTUs must therefore be at least somewhat functionally equivalent with respect to the traits examined. Functional equivalency in the gut microbiome could offer an adaptive advantage to the host by improving community resilience in response to disturbance [@Ley2006b]. Lastly, mean compositional differences among infants born by C-section were, on average, notably greater both in terms of OTU-based and trait-based dissimilarity ((ref:figInterbabyDissTreat)). Such differences could arise if the taxa to which C-section infants are initially exposed are more taxonomically and functionally variable than the taxa to which vaginally delivered infants are exposed. 
  
\newline

#Conclusions

\newline

As in the ecological studies of macroorganisms, trait-based analysis of gut microbiome succession offers insights into the mechanisms of community assembly, such as dispersal limitation and ecological filtering, and the balance between stochastic and deterministic forces. The stabilization of trait-based community composition over time ((ref:figIntrababyDiss)), and the drop in variance of community trait values for most traits over time ((ref:figCWMsVariance)), both suggests that succession is at least partially functionally deterministic, with early dynamics possibly reflecting stochastic colonization during the birthing process, followed by the gradual colonization and enrichment of taxa better adapted for the mature gut environment. Steady directional turnover in OTUs over succession, even after trait-based composition had stabilized, underscores that changes in OTU-based composition need not imply changes in trait-based composition [@Strickland2009; @Shetty2017b], but also that there are surely aspects of community assembly that cannot yet be understood by using only the traits used in this study. Future work should expand the number of traits considered. Moreover, because our study is observational, we cannot distinguish between an OTU that fails to disperse to a potential host and an OTU that arrives but fails to establish, so future research should also explore the relationship between OTU arrival and detection in fecal samples to better disentangle dispersal limitation and niched-based differences among taxa.

`r indent` Comparisons of trait-based patterns between cohorts of infants are an opportunity to understand the effects of specific events (e.g., delivery mode, antibiotic exposure), but more generally, are natural experiments that can reveal how gut communities respond to, and recover from, systematic disturbances. In our analysis, for example, delivery mode resulted in sustained differences in community composition, indicating that priority effects can play an important role in gut community assembly [@Sprockett2018; @Fukami2011b], a result that likely extends to other types of disturbance during early life. Similarly, repeated antibiotics treatments led to sustained differences in trait-based community composition ((ref:figCWMsTreat)), suggesting that gut communities are not infinitely functionally resistant and/or that tradeoffs exist between antibiotic resistance and other traits [@Allison2008]. Understanding trait-based differences between other cohorts, such as healthy vs. diseased [@Sekirov2010b], or on and off specific diets [@David2014b], could provide insight into additional factors shaping gut microbiome community assembly. For example, the unhealthy, dysbiotic gut may have a higher prevalence of microaerobic and biofilm-forming species [@Hughes2017; @Srivastava2017], a difference could be detected by trait-based analyses. Trait-based approaches, which link organismal structures to key ecological functions, are poised to advance our mechanistic understanding of the gut microbiome, and their usefulness will only increase as we improve our knowledge of how traits mediate microbial interactions and acquire trait data for a wider range of taxa.

\newline

#Author contributions

JG, AS and EL conceived the study, JG developed methods, analyzed data, and wrote the manuscript, and all authors discussed analysis and revised the manuscript.

#Acknowledgments

This work was supported in part by Michigan State University through computational resources provided by the Institute for Cyber-Enabled Research. AS acknowledges support from the National Science Foundation under Grant No DEB#1749544. JG was supported by the Michigan State University Foundation funding to EL.

`r pagebreak`

```{r --Tables and Figures--, include = FALSE} 
``` 


```{r tabTraitSources, include = FALSE}

ref(type = 'table', name = 'tabTraitSources')

#note how I add a special character to the legend with \\\\'a
tmp <- kable(tabTraitSources, format = 'latex', escape = F, booktabs = T, align = c('l','l','l'), linesep = "") %>%
  kable_styling() %>%
  row_spec(0, bold = T) %>% 
  landscape() %>%
  footnote(threeparttable = TRUE, 
           general = "Sources of trait data gathered in this study. IgA: Immunoglobulin A; BacDive: Bacterial Diversity Metadatabase (Sohngen et al. 2014). IJSEM: International Journal of Systematic and Evolutionary Microbiology. JGI: Joint Genome Institute (Also see: Barber\\\\'an et al. 2017); NCBI: National Center for Biotechnology Information; rrnDB: the ribosomal RNA operon copy number database (Stoddard et al. 2015). For additional details see Methods.", 
           general_title = paste0("Table ", refs$tabTraitSources, "."), 
           title_format = "bold", escape = F)

kable_as_image(tmp, filename = paste0(image_dir,'tabTraitSources'), density = 500)

```

(ref:tabTraitSources) Table `r refs$tabTraitSources`

`r pagebreak` ![](images\\tabTraitSources.png)

```{r tabThresholds, include = FALSE}

ref(type = 'table', name = 'tabThresholds')

#i cannot explain it, but the rmarkdown doesn't produce this image properly. has to be done manually
if (FALSE) {
  tmp <- kable(tabThresholds, format = 'latex', escape = F, booktabs = T, align = c('l','r'), linesep = "") %>%
    kable_styling() %>%
    row_spec(0, bold = T) %>% 
    footnote(threeparttable = TRUE, 
             general = "Percent sequence dissimilarity (i.e., phylogenetic distance) in the 16S rRNA V4 region at which statistical support for trait conservatism disappears (see Methods and Fig. S4)", 
             general_title = paste0("Table ", refs$tabThresholds, "."), 
             title_format = "bold")
  
  kable_as_image(tmp, filename = paste0(image_dir,'tabThresholds'), density = 500)

}

```

(ref:tabThresholds) Figure `r refs$tabThresholds`

`r pagebreak` ![](images\\tabThresholds.png){width=4.5in}

```{r figTaxOverTime, include = FALSE}

#it is necessary to do this because the normal plot_it() function doesn't work for grid objects
png(filename = paste0(image_dir, "figTaxOverTime", '.png'), units = 'in', type = "cairo", width = 6.5, height = 6.5, res = 300)
grid::grid.newpage()
grid::grid.draw(figTaxOverTime)
while(dev.cur() != 1) dev.off()

#log ref
ref(type = 'figure', name = 'figTaxOverTime')

```

(ref:figTaxOverTime) Figure `r refs$figTaxOverTime`

`r pagebreak` ![](images\\figTaxOverTime.png)

**(ref:figTaxOverTime):** OTU and family-level abundance patterns over time. Colors reflect successional status; taxa were categorized as early/late successional if their abundances across all infants trended significantly negative/positive (p < 0.05) over time based on linear regression; OTUs that did not trend over time were grouped into the "mid-successional or no trend" category. Left: percent total abundance of OTUs from each successional category over time. Right: proportions of total abundance of the top five most-abundant families in each successional group.

```{r figCWMs, include = FALSE}

plot_it(figCWMs, name = 'figCWMs', dir = image_dir, width = 6.5, height = 7.5)

```

(ref:figCWMs) Figure `r refs$figCWMs`

`r pagebreak` ![](images\\figCWMs.png)

**(ref:figCWMs):** Abundance-weighted trait patterns over time. Red circles show average abundance-weighted trait means of samples in that month. N is equal to the number of samples, and ranges from `r n_cwms_min` to `r n_cwms_max`. Vertical red lines show 95 percent confidence intervals. Black trendlines were fit using generalized additive models.

```{r figCWMsTreat, include = FALSE}

plot_it(figCWMsTreat, name = 'figCWMsTreat', dir = image_dir, width = 6.5, height = 7)

```

(ref:figCWMsTreat) Figure `r refs$figCWMsTreat`

`r pagebreak` ![](images\\figCWMsTreat.png)

**(ref:figCWMsTreat):** Abundance-weighted trait patterns over time, separated by delivery mode and antibiotic regime. Filled circles show average abundance-weighted trait means of samples within six-month periods for each cohort of infants. N is equal to the number of samples of each treatment within each six-month period; there were between `r n_cs_min` to `r n_cs_max` total samples from `r N_cs` infants delivered by C-section (who received little to no antibiotics), `r n_ab_min` to `r n_ab_max` total samples from `r N_ab` infants who were treated for at least 50 days, and `r n_cntrl_min` to `r n_cntrl_max` total samples from `r N_control` "Control" infants that were delivered vaginally and received no antibiotics. Vertical lines show 95 percent confidence intervals. Asterisks denote significance based on Welch t-tests performed between each treatment group and the control group, and are colored to correspond to treatment (\*: adjusted p < 0.05; \*\*: adjusted p < 0.01; \*\*\*: adjusted p < 0.001).

```{r figIntrababyDiss, include = FALSE}

plot_it(figIntrababyDiss, name = 'figIntrababyDiss', dir = image_dir, width = 5.5, height = 5.5)

```

(ref:figIntrababyDiss) Figure `r refs$figIntrababyDiss`

`r pagebreak` ![](images\\figIntrababyDiss.png)

**(ref:figIntrababyDiss):** Taxonomic and trait-based rates of community change within infants over time. Filled black circles show mean pairwise dissimilarities between microbiome samples taken from the same infants, grouped into six-month periods. OTU-based differences were calculated using Bray-Curtis dissimilarity. Trait-based dissimilarities were calculated using multidimensional Euclidean distance (Petchey and Gaston 2002) after scaling each trait so that it has the potential to contribute equally. Thus, N equals `r nrow(meta)`, the number of infants. Vertical lines show 95 percent confidence intervals. Unfilled circles in the bottom two panels also reflect trait-based Euclidean distances between pairs of samples, but with a version of the data where trait values were randomly shuffled among OTUs. Hence, the significant differences between filled and unfilled circles suggest that rates of trait-based convergence were faster than predicted by a trait-agnostic null model of community change over time. Asterisks denote significance between observed and null data, based on Welch t-tests (\*: adjusted p < 0.05; \*\*: adjusted p < 0.01; \*\*\*: adjusted p < 0.001).

```{r figInterbabyDiss, include = FALSE}

plot_it(figInterbabyDiss, name = 'figInterbabyDiss', dir = image_dir, width = 5.5, height = 4)

```

(ref:figInterbabyDiss) Figure `r refs$figInterbabyDiss`

`r pagebreak` ![](images\\figInterbabyDiss.png)

**(ref:figInterbabyDiss):** Mean dissimilarity in gut microbiome community compositions across infants over early development. Dissimilarities were calculated for all sample pairs within six-month periods, excluding pairs of samples from the same infants. OTU-based differences were calculated using Bray-Curtis dissimilarity. Trait-based dissimilarities were calculated using multidimensional Euclidean distance (Petchey and Gaston 2002) after scaling each trait so that it has the potential to contribute equally. Unfilled circles in the bottom two panels also reflect trait-based Euclidean distances between pairs of samples, but with a version of the data where trait values were randomly shuffled among OTUs. Hence, the significant differences between filled and unfilled circles suggest that rates of trait-based convergence across infant microbiomes was faster than predicted by a trait-agnostic null model of community change over time. Asterisks denote significance between observed and null data, based on Welch t-tests (\*: p < 0.05; \*\*: p < 0.01; \*\*\*: p < 0.001)

```{r --Supplementary--, include = FALSE} 
``` 

```{r tabHSP, include = FALSE}

ref(type = 'supplementary table', name = 'tabHSP')

tabHSP <- kable(tabHSP, format = 'latex', escape = F, booktabs = T, align = c('l','r','r','r'), linesep = "", digits = 2) %>%
  kable_styling() %>%
  row_spec(0, bold = T) %>% 
  footnote(threeparttable = TRUE, 
           general = "Pearson correlations coefficients among different methods of hidden state character prediction. IC: Independent Contrasts; SBT: Subtree Averaging; SCP: Weighted Squared-change Parsimony. Traits that were not amenable to hidden state prediction were omitted due to insufficient data.", 
           general_title = paste0("Table S", refs$tabHSP, "."), 
           title_format = "bold")

kable_as_image(tabHSP, filename = paste0(image_dir,'tabHSP'), density = 500)

```

(ref:tabHSP) Figure S`r refs$tabHSP`

`r pagebreak` ![](images\\tabHSP.png){width=4.5in}

```{r figSampling, include = FALSE}

plot_it(figSampling, type = 'supplementary figure', name = 'figSampling', dir = image_dir, width = 6.5, height = 5.5)

```

(ref:figSampling) Figure S`r refs$figSampling`

`r pagebreak` ![](images\\figSampling.png)

**(ref:figSampling):** Sampling times, sampling windows, and delivery modes of infants used in this study. Each horizontal line is a subject; each filled circle is a sample.

```{r figTreeTraits, include = FALSE}

plot_it(figTreeTraits, type = 'supplementary figure', name = 'figTreeTraits', dir = image_dir, width = 6.5, height = 8.5)

```

(ref:figTreeTraits) Figure S`r refs$figTreeTraits`

`r pagebreak` ![](images\\figTreeTraits.png)

**(ref:figTreeTraits):** Left: a phylogenetic tree constructed from a distance matrix of sequence similarity in the 16S rRNA V4 region. The tips of the tree comprise the 3311 OTUs in this study (before rarefaction) and the 13900 OTUs from the 132 release of the Living Tree Project [@Munoz2011]. The color of the horizontal lines to the right of the tree reflects the OTU source. Right: a jittered scatterplot illustrating the number and phylogenetic extent of individual trait observations compiled from databases and the literature (ref:tabTraitSources).

```{r figThresholds, include = FALSE}

plot_it(figThresholds, type = 'supplementary figure', name = 'figThresholds', dir = image_dir, width = 7, height = 8.5)

```

(ref:figThresholds) Figure S`r refs$figThresholds`

`r pagebreak` ![](images\\figThresholds.png)

**(ref:figThresholds):** Quantifying the limits of trait-based phylogenetic inference for each trait. Each filled black circle shows the mean trait-based difference among 1000 randomly selected pairs of taxa that fall within 0.5 percent bins of increasing sequence dissimilarity in the 16S rRNA V4 region. Horizontal dotted lines show null expectations for each trait, equal to the mean trait-based difference across all pairs of taxa with at least 10 percent sequence dissimilarity. The size (i.e., the number of tips) of the phylogeny used for each trait varied, and was equal to the number of OTUs with known trait values. Hence, the number of pairwise comparisons differed by trait, ranging from `r tree_Npairs$n[1]` (`r tree_Npairs$trait[1]`) to `r tree_Npairs$n[2]` (`r tree_Npairs$trait[2]`). Colored trendlines show the best-fit models for each trait, as determined by AIC, selected from a suite of five basic models (Null: $y \sim 1$; Linear regression: $y\sim x$; Asymptotic regression: $y \sim a(1 - e^{(-e^bx)})$, where $a$ and $b$ were determined using a self-starting nonlinear least squares approach, and the model fit was constrained to pass through the origin; Logistic regression: $y \sim \frac {a}{1 + e^(\frac{b-x}{c}))}$, where $a$, $b$, and $c$ were determined using a self-starting nonlinear least squares approach. Vertical dashed lines mark the maximum phylogenetic distances at which unknown trait values were inferred for each trait in this study, and were calculated as the point at which the best-fit model reached 90 percent of the null expectation of pairwise trait dissimilarity.

```{r figTraitCoverage, include = FALSE}

plot_it(figTraitCoverage, type = 'supplementary figure', name = 'figTraitCoverage', dir = image_dir, width = 6.5, height = 5.5)

```

(ref:figTraitCoverage) Figure S`r refs$figTraitCoverage`

`r pagebreak` ![](images\\figTraitCoverage.png)

**(ref:figTraitCoverage):** Proportions of sequences in this study with trait data that (1) were drawn from the literature, (2) were inferred using hidden state prediction (see Methods), or (3) could not be inferred and were therefore dropped from analyses. Trait data could not be inferred when trait values did not exhibit sufficient phylogenetic signal ((ref:figThresholds)) and/or when there were insufficient trait data on closely related OTUs ((ref:figTreeTraits)).

```{r figCWMsSuccGroup, include = FALSE}

plot_it(figCWMsSuccGroup, type = 'supplementary figure', name = 'figCWMsSuccGroup', dir = image_dir, width = 7.5, height = 7)

```

(ref:figCWMsSuccGroup) Figure S`r refs$figCWMsSuccGroup`

`r pagebreak` ![](images\\figCWMsSuccGroup.png)

**(ref:figCWMsSuccGroup):** Mean trait values of OTUs from early (N = 139), middle or no-trend (i.e., "Other;" N = 209), and late (N = 393) successional stages. Unlike elsewhere, these calculations are not weighted by abundance. Vertical bars show 95 percent confidence intervals. OTUs were assigned to successional groups according to whether they exhibited significant positive or negative trends in overall abundance over succession based on linear models. Lower case letters (a/b/c) denote significance (p < 0.05) among successional categories for each trait, calculated using Welch t-tests for each comparison.

```{r figTraitCorr, include = FALSE}

plot_it(figTraitCorr, type = 'supplementary figure', name = 'figTraitCorr', dir = image_dir, width = 6.5, height = 5)

```

(ref:figTraitCorr) Figure S`r refs$figTraitCorr`

`r pagebreak` ![](images\\figTraitCorr.png)

**(ref:figTraitCorr):** Pearson correlation coefficients among traits. Asterisks denote significance between trait values among the `r nrow(traits_wide)` OTUs in this study with at least one associated trait datum (\*: adjusted p < 0.05; \*\*: adjusted p < 0.01; \*\*\*: adjusted p < 0.001.

```{r figSporulatorsTime, include = FALSE}

plot_it(figSporulatorsTime, type = 'supplementary figure', name = 'figSporulatorsTime', dir = image_dir, width = 7.5, height = 7)

```

(ref:figSporulatorsTime) Figure S`r refs$figSporulatorsTime`

`r pagebreak` ![](images\\figSporulatorsTime.png)

**(ref:figSporulatorsTime):** Numbers of OTUs with sporulation scores greater or less than 0.1. Each point represents one sample from one infant. Smooth trendlines were fitted using a generalized additive model. The rise in taxa with sporulation scores > 0.1 over time illustrates that it is not merely an increase in the abundance of preexisting taxa driving the observed increase in abundance-weighted sporulation score over time (Figure 2), but also an increase in OTU richness.


```{r figCWMsVariance, include = FALSE}

plot_it(figCWMsVariance, type = 'supplementary figure', name = 'figCWMsVariance', dir = image_dir, width = 6.5, height = 7)

```

(ref:figCWMsVariance) Figure S`r refs$figCWMsVariance`

`r pagebreak` ![](images\\figCWMsVariance.png)

**(ref:figCWMsVariance):** Community trait variance of infant gut microbiomes over time. Values reflect the variance of microbial trait values in gut microbiome samples over time. Values are scaled for each trait to enable across-trait comparison. Circles show average variance of samples in that month. N is equal to the number of samples, and ranges from `r n_cwms_min` to `r n_cwms_max`. Vertical lines show 95 percent confidence intervals. Trendlines show shifts in community trait variance over time, calculated using generalized additive models.


```{r figInterbabyDissTreat, include = FALSE}

plot_it(figInterbabyDissTreat, type = 'supplementary figure', name = 'figInterbabyDissTreat', dir = image_dir, width = 6.5, height = 4)

```

(ref:figInterbabyDissTreat) Figure S`r refs$figInterbabyDissTreat`

`r pagebreak` ![](images\\figInterbabyDissTreat.png)

**(ref:figInterbabyDissTreat):** Mean dissimilarity in gut microbiome community composition across infants over early development, after subsetting infants by antibiotic history and delivery mode. Mean dissimilarities were calculated for all sample pairs within six-month periods within treatment groups, excluding pairs of samples from the same infants. OTU-based differences were calculated using Bray-Curtis dissimilarity. Trait-based dissimilarities were calculated using multidimensional Euclidean distance after scaling each trait so that it has the potential to contribute equally. N equals the number of all possible sample pairs within each treatment; within each six-month period, there were between `r n_cs_min` to `r n_cs_max` total samples from `r N_cs` infants delivered by C-section, `r n_ab_min` to `r n_ab_max` total samples from `r N_ab` infants that received at least 50 days of antibiotic treatment, and `r n_cntrl_min` to `r n_cntrl_max` total samples from `r N_control` "Control" infants that were delivered vaginally and did not receive antibiotics. Due to the large number of pairwise comparisons, and therefore sample sizes, both treatment groups are almost always significantly different from the "Control" infant group during all time periods.

`r pagebreak` Works Cited
